### ISSUE TITLE
[Bug]: traverse does not calculate scope in object deconstructor in catch clause

### ISSUE THREAD

**Issue by overlookmotel** [babel/babel#14024]
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
const {transformSync} = require('@babel/core');

const codeIn = `
  try {
    throw new Error("oops");
  } catch ( {
    message,
    f = () => message
  } ) {}
`;

const {code} = transformSync(codeIn, {
  plugins: [
    () => ({
      visitor: {
        Identifier(path) {
          if (!path.parentPath.isFunction()) return;
          if (!path.scope.getBinding('message')) path.node.name += '_unbound';
        }
      }
    })
  ]
});

console.log(code);
// ... f = () => message_unbound ...
```

### Configuration

n/a

### Current and expected behavior

As in example above:

```js
try {
  throw new Error("oops");
} catch ( {
  message,
  f = () => message
} ) {}
```

This code is a little odd, but legal!

`message` in `() => message` should be bound to the deconstructed property `message` of the error object.

However, `path.scope.getBinding('message')` returns `undefined`.

### Environment

```
System:
  OS: macOS 10.15.7
Binaries:
  Node: 16.13.1 - ~/.nvm/versions/node/v16.13.1/bin/node
  npm: 8.1.2 - ~/.nvm/versions/node/v16.13.1/bin/npm
npmPackages:
  @babel/core: ^7.16.0 => 7.16.0
  @babel/generator: ^7.16.0 => 7.16.0
  @babel/helper-module-transforms: ^7.16.0 => 7.16.0
  @babel/helper-plugin-utils: ^7.14.5 => 7.14.5
  @babel/parser: ^7.16.4 => 7.16.4
  @babel/plugin-transform-arrow-functions: ^7.16.0 => 7.16.0
  @babel/plugin-transform-modules-commonjs: ^7.16.0 => 7.16.0
  @babel/plugin-transform-react-jsx: ^7.16.0 => 7.16.0
  @babel/plugin-transform-strict-mode: ^7.16.0 => 7.16.0
  @babel/register: ^7.16.0 => 7.16.0
  @babel/traverse: ^7.16.3 => 7.16.3
  @babel/types: ^7.16.0 => 7.16.0
  babel-jest: ^27.4.2 => 27.4.2
  babel-plugin-dynamic-import-node: ^2.3.3 => 2.3.3
  eslint: ^7.32.0 => 7.32.0
  jest: ^27.4.3 => 27.4.3
```

### Additional context

I would be happy to work on a fix. Any pointers on where to look in the codebase would be appreciated.

**Comment by lightmare** [comment]
The culprit seems to be this condition:
https://github.com/babel/babel/blob/2d989a983dc13c7682d65faa075c8a26d2f5b991/packages/babel-traverse/src/scope/index.ts#L1131-L1136

I tried adding a check for `!scope.path.isCatchClause()`, but that broke this test:
https://github.com/babel/babel/blob/2d989a983dc13c7682d65faa075c8a26d2f5b991/packages/babel-traverse/test/fixtures/rename/catch-clause/input.js#L1-L5

I think to fix both a new `BindingKind` for catch params is needed.

**Comment by JLHwung** [comment]
We can register `"param"` binding for catch clause

https://github.com/babel/babel/blob/a943f576d1f7181abc593f97b174fdcf8a6c421d/packages/babel-traverse/src/scope/index.ts#L334

And for declarations within the catch body, they will be registered in BlockStatement.

I am not sure why it is registered as `let` and it seems that we don't have tests covering that, too.

**Comment by overlookmotel** [comment]
Thanks for swift replies.

@JLHwung 

> And for declarations within the catch body, they will be registered in BlockStatement.

This would be inconsistent with declarations in a function body, which are registered to the FunctionDeclaration / FunctionExpression.

e.g.: `function f() { let x; return x; }` 

If `path` is the identifier `x` in `return x`, `path.getBinding('x').scope.path.type === 'FunctionDeclaration'`.

Personally, I think that behavior is unintuitive - I would have expected it to be `'BlockStatement'`.

Can you advise if that is desired behavior? If not, I'll make a PR to fix that before fixing behavior of params in CatchClause - so the two are consistent.

**Comment by JLHwung** [comment]
@overlookmotel You are right. We don't generate scope when BlockStatement is immediately nested in Function/CatchClause,

https://github.com/babel/babel/blob/94af0e5c629b63fa11512b2ea8a8ac81d11a9db5/packages/babel-types/src/validators/isScope.ts#L14-L18

so declarations within the catch body is registered in CatchClause.

> I would have expected it to be 'BlockStatement'

I agree with you. I guess merging BlockStatement with FunctionDeclaration is for performance: So we only generate one scope for function, instead of one for param / id and the other for block body.
